### PR TITLE
feat(relay): Add relay dialogs

### PR DESCRIPTION
**Pull request by priscilawebdev**
**Description:**

Added Relay Edit and Add Dialogs:

image

image

image

#sync-getsentry

### CODE REVIEW

File: src/sentry/static/sentry/app/components/forms/inputField.tsx [hunk 1]
```diff
@@ -4,7 +4,7 @@ import React from 'react';
 import FormField from 'app/components/forms/formField';
 
 type InputFieldProps = FormField['props'] & {
-  placeholder: string;
+  placeholder?: string;
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/forms/inputField.tsx:7]: What's the reason for making this optional?

**Review by billyvg — COMMENTED** [review]
Can we use `globalModal` here instead of re-implementing? Is there a need to couple the modal with the form? I think they can remain 2 separate components.

File: src/sentry/static/sentry/app/views/settings/organizationRelays/dialogs/form.tsx [hunk 2]
```diff
@@ -0,0 +1,76 @@
+import React from 'react';
+import styled from '@emotion/styled';
+
+import {t} from 'app/locale';
+import TextField from 'app/components/forms/textField';
+import TextareaField from 'app/components/forms/textareaField';
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/settings/organizationRelays/dialogs/form.tsx:6]: Let's use the form components in `app/views/settings/components/form`

File: src/sentry/static/sentry/app/views/settings/organizationRelays/dialogs/form.tsx [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        onChange={(value: string) => {
+          onChange('name', value);
+        }}
+        value={values.name}
+        onBlur={onValidate('name')}
+        error={errors.name}
+        disabled={disables.name}
+      />
+    </FormField>
+    <FormField label={t('Key')} isFullWidth>
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/settings/organizationRelays/dialogs/form.tsx:36]: The `FormField` in `app/views/settings/components/form` has a prop for these kind of layout: https://storybook.getsentry.net/branches/master/index.html?path=/story/forms-form--default

File: src/sentry/static/sentry/app/components/forms/inputField.tsx (hunk 1 above)
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/components/forms/inputField.tsx:7]: Sometimes only the field title (Label) is enough

File: src/sentry/static/sentry/app/views/settings/organizationRelays/dialogs/form.tsx (hunk 2 above)
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/settings/organizationRelays/dialogs/form.tsx:6]: can we please add  "this component is deprecated" inside `app/components/forms/* `?

**Comment by priscilawebdev** [comment]
@billyvg sure! I updated the PR 😉 Thank you for the nice feedback

**Comment by billyvg** [comment]
@priscilawebdev Was there a weird rebase? Lots of other files are being included in the diff

File: src/sentry/api/endpoints/organization_relay_keys.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+            raise APIForbidden("Relay configuration is not enabled for this organization.")
+
+    def get(self, request, organization):
+        self.check_feature_enabled(request, organization)
+
+        relay_keys = organization.get_option(ORG_RELAYS_OPTION, [])
+        public_keys = [key_info.get("public_key") for key_info in relay_keys]
+        usage = _get_keys_usage(public_keys)
+        ret_val = _add_usage_info(relay_keys, usage)
+        return Response(_to_camel(ret_val), status=status.HTTP_200_OK)
```
Comment (markstory) [inline src/sentry/api/endpoints/organization_relay_keys.py:39]: Generally we would use rest_framework serializers, would that be something we could do here? It would give you a place to put your _ prefixed functions.

File: src/sentry/api/endpoints/organization_relay_keys.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+    def put(self, request, organization, public_key):
+        self.check_feature_enabled(request, organization)
+        data = _to_snake(request.json_body)
+        if data is None:
+            return Response(_error("missing request body"), status=status.HTTP_400_BAD_REQUEST)
```
Comment (markstory) [inline src/sentry/api/endpoints/organization_relay_keys.py:45]: You could use rest_framework.exceptions.ParseError here to match other API endpoints, and remove the need for the `_error` function.

File: src/sentry/api/endpoints/organization_relay_keys.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+                # the key was already used change the name/description
+                existing_key["name"] = data.get("name")
+                existing_key["description"] = data.get("description")
+                ret_val = existing_key
+                status_code = status.HTTP_200_OK
+                break
+        else:
+            data["public_key"] = public_key
+            data["created"] = datetime.utcnow()
+            relay_keys.append(data)
```
Comment (markstory) [inline src/sentry/api/endpoints/organization_relay_keys.py:60]: Are we confident at this point that the SSH key is valid?

File: src/sentry/api/endpoints/organization_relay_keys.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+        organization.update_option(ORG_RELAYS_OPTION, relay_keys)
+        return Response(_to_camel(ret_val), status=status_code)
+
+    def delete(self, request, organization, public_key):
+        self.check_feature_enabled(request, organization)
+        old_keys = organization.get_option(ORG_RELAYS_OPTION, [])
+        new_keys = [key_info for key_info in old_keys if key_info.get("public_key") != public_key]
+
+        if len(old_keys) == len(new_keys):
+            return Response(_error("invalid_relay_key"), status=status.HTTP_404_NOT_FOUND)
```
Comment (markstory) [inline src/sentry/api/endpoints/organization_relay_keys.py:75]: sentry.api.exceptions:ResourceDoesNotExist could work here.

File: src/sentry/lang/java/plugin.py [hunk 8]
```diff
@@ -32,16 +29,6 @@ def handles_frame(self, frame, stacktrace_info):
...
 
-    def preprocess_frame(self, processable_frame):
-        processable_frame.set_cache_key_from_values(
-            (
-                FRAME_CACHE_VERSION,
-                processable_frame.frame["module"],
-                processable_frame.frame["function"],
-            )
-            + tuple(sorted(map(six.text_type, self.images)))
-        )
```
Comment (markstory) [inline src/sentry/lang/java/plugin.py:43]: Is this related to symbolic updates?

File: src/sentry/static/sentry/app/views/releasesV2/list/crashFree.tsx [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+import space from 'app/styles/space';
+import {IconFire, IconWarning, IconCheckmark} from 'app/icons';
+
+import {displayCrashFreePercent} from '../utils';
+
+const CRASH_FREE_DANGER_THRESHOLD = 98;
+const CRASH_FREE_WARNING_THRESHOLD = 99.5;
+
+const getIcon = (percent: number) => {
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/releasesV2/list/crashFree.tsx:12]: I think your rebase has gone awry here.

**Comment by priscilawebdev** [comment]
@markstory @billyvg it might be because I cherry-pick some @RaduW commits... let me rebase it again :)

File: src/sentry/static/sentry/app/views/releasesV2/list/crashFree.tsx (hunk 9 above)
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/releasesV2/list/crashFree.tsx]: yes...I just fixed it... I'm sorry

**Review by doralchan — CHANGES_REQUESTED** [review]
Design Changes - https://www.figma.com/file/et5ZVSG4OTjfUkRAC0bCCb/Proxy-Service-Relay?node-id=0%3A1

File: src/sentry/api/endpoints/organization_relay_keys.py (hunk 4 above)
Comment (priscilawebdev) [reply src/sentry/api/endpoints/organization_relay_keys.py:39]: @RaduW I believe you have some feedbacks 😉

File: src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx [hunk 10]
```diff
@@ -10,69 +11,76 @@ import {Organization} from 'app/types';
...
+      this.setState(prevState => ({
+        relays: prevState.relays.filter(relay => relay.publicKey !== publicKey),
+      }));
+    } catch {
+      addErrorMessage('An unknown error occurred while deleting relay public key');
+    }
+  };
+
+  handleSave = async (data: Parameters<Add['props']['onSave']>[0]) => {
+    try {
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx:53]: Do you need to indicate a loading state before sending the request?

File: src/sentry/static/sentry/app/views/settings/organizationRelays/dialogs/dialogManager.tsx [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+  handleSave = () => {
+    const {onSave, closeModal} = this.props;
+
+    if (onSave) {
+      onSave(this.state.values);
+    }
+
+    closeModal();
+  };
```
[the remark this answers is not in the document]
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/settings/organizationRelays/dialogs/dialogManager.tsx:82]: I believe you saw it in the data Scrubbing.. I am planning to do it as soon :)

File: src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx (hunk 10 above)
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx:53]: It's good to have..let me add it

**Review by doralchan — APPROVED** [review]
Looks good design wise

File: src/sentry/static/sentry/app/views/settings/organizationRelays/modals/handleXhrErrorResponse.tsx [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+type Error = {
+  type:
+    | 'unknown'
+    | 'bad-structure'
+    | 'missing-name'
+    | 'empty-name'
+    | 'missing-key'
+    | 'invalid-key';
+  message: string;
+};
```
Comment (RaduW) [inline src/sentry/static/sentry/app/views/settings/organizationRelays/modals/handleXhrErrorResponse.tsx:12]: now there is also a "duplicate key" error

File: src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx [hunk 13]
```diff
@@ -34,154 +38,237 @@ type State = AsyncComponent['state'] & {
...
   getDefaultState() {
     return {
       ...super.getDefaultState(),
-      relays: relaysMock,
+      relays: [],
+    };
+  }
+
+  getEndpoints(): ReturnType<AsyncComponent['getEndpoints']> {
+    return [['data', `/organizations/${this.props.organization.slug}/`]];
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx:59]: Do you need to re-fetch the organization? You could use `withOrganization()` to get the current organization data.

File: src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx [hunk 14]
```diff
@@ -34,154 +38,237 @@ type State = AsyncComponent['state'] & {
...
-            <Button onClick={this.handleAdd} priority="primary">
-              {t('Add Relay')}
-            </Button>
-          </PanelAction>
-        </Panel>
+          }
+        />
+        <TextBlock>
+          {tct(
+            `Relay is a relay service built by Sentry. You can run this on-premise for your SDKs or server to customize data scrubbing, buffering retries and more. You can generate relay keys for access. For more on how to set this up, read the [link:docs].`,
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx:158]:
```suggestion
            `Relay is a forwarding service built by Sentry. You can run this on-premise for your SDKs or server to customize data scrubbing, buffering retries and more. You can generate relay keys for access. For more on how to set this up, read the [link:docs].`,
```

'Relay is a relay' is awkward would 'proxy' or 'forwarding' work?

File: src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx [hunk 15]
```diff
@@ -34,154 +38,237 @@ type State = AsyncComponent['state'] & {
...
+              headers={[t('Display Name'), t('Relay Key'), t('Date Created'), '']}
+              className={css`
+                grid-template-columns: repeat(3, auto) max-content;
+                > * {
+                  padding: ${space(1)};
+                  @media (min-width: ${theme.breakpoints[0]}) {
+                    padding: ${space(1)} ${space(2)};
+                  }
+                }
+              `}
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx:178]: This could be a styled component instead of inline css.

File: src/sentry/static/sentry/app/views/settings/organizationRelays/modals/handleXhrErrorResponse.tsx (hunk 12 above)
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/settings/organizationRelays/modals/handleXhrErrorResponse.tsx:12]: I'll add it 😉

File: src/sentry/static/sentry/app/views/settings/organizationRelays/modals/modalManager.tsx [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+  constructor(props: P) {
+    super(props);
+    this.handleSave = this.handleSave.bind(this);
+    this.clearError = this.clearError.bind(this);
+    this.convertErrorXhrResponse = this.convertErrorXhrResponse.bind(this);
+    this.handleChange = this.handleChange.bind(this);
+    this.handleValidateForm = this.handleValidateForm.bind(this);
+    this.handleValidate = this.handleValidate.bind(this);
+    this.handleValidateKey = this.handleValidateKey.bind(this);
+    this.setValidForm = this.setValidForm.bind(this);
```
[the remark this answers is not in the document]
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/settings/organizationRelays/modals/modalManager.tsx:47]: good catch... I removed the bind from the funcs that return` () =>`

File: src/sentry/static/sentry/app/views/settings/organizationRelays/modals/handleXhrErrorResponse.tsx [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+  const errorMessage = error.responseJSON?.trustedRelays[0];
+
+  if (!errorMessage) {
+    return {
+      type: 'unknown',
+      message: t('An unknown error occurred while saving relay public key'),
+    };
+  }
+
+  if (errorMessage === 'Bad structure received for Trusted Relays') {
```
[the remark this answers is not in the document]
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/settings/organizationRelays/modals/handleXhrErrorResponse.tsx:30]: There was an offline discussion about this. The back-end team decided to follow the current standard, returning a more readable answer for the user in the request-response. If we change this now, we should also update the other APIs. We will discuss it during the TSC meeting

File: src/sentry/static/sentry/app/views/settings/organizationRelays/modals/handleXhrErrorResponse.tsx [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+  };
+};
+
+function handleError(error: XhrError): Error {
+  const errorMessage = error.responseJSON?.trustedRelays[0];
+
+  if (!errorMessage) {
+    return {
+      type: 'unknown',
+      message: t('An unknown error occurred while saving relay public key'),
```
[the remark this answers is not in the document]
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/settings/organizationRelays/modals/handleXhrErrorResponse.tsx:26]: we have more specific error messages, for example, if the user enters an invalid key, the error "invalid key" will be shown. 

I would also like to return a more abstract error message, as it can be anything ... it can be a server problem, etc...not necessarily an error generated by the user

File: src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx (hunk 14 above)
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx:158]: There was a discussion about it here: https://www.figma.com/file/et5ZVSG4OTjfUkRAC0bCCb/Proxy-Service-Relay?node-id=0%3A1

In final we decided to keep "relay" here, because It is not really a proxy but a relay... It  relays data to Sentry and processes it along the way...

File: src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx (hunk 15 above)
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx:178]: This is not possible, because the component "PanelTable" is a func component and not a styled component. It accepts only a className.

As we discussed offline, I would love to use the css provided by emotion in the prop className. Example: className={css`...`}, but it doesn't work...it has another return type, etc.... so I decided to use ClassNames 😉

File: src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx [hunk 19]
```diff
@@ -34,154 +37,224 @@ type State = AsyncComponent['state'] & {
...
+
+      try {
+        const response = await this.api.requestPromise(
+          `/organizations/${this.props.organization.slug}/`,
+          {
+            method: 'PUT',
+            data: {trustedRelays},
+          }
+        );
+        addSuccessMessage('Successfully deleted relay public key');
```
Comment (jan-auer) [inline src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx:77]: _nit_: Please capitalize _"Relay"_ here and in all other instances. _"public key"_ can stay lower case, in my opinion.

**Review by jan-auer — APPROVED** [review]
Thank you! Tests on staging were successful, and I was able to ingest an event through a custom Relay.

Approving for initial feature-flagged customer testing. Before we EA this, we can come back and update the introduction text with a mini tutorial for setting Relay up. We're currently working on updated documentation, which we can incorporate then.

Note that I did not review the code, leaving this to the Pros.

File: src/sentry/static/sentry/app/views/settings/organizationRelays/modals/handleXhrErrorResponse.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/settings/organizationRelays/modals/handleXhrErrorResponse.tsx:31]: I was also thinking of displaying only the server error in the user interface, but I sometimes display the error message in specific locations according to the message. Example: "Invalid Relay key" should be displayed along with the relay field .. for this reason, I prefer to leave it like this

File: src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx [hunk 20]
```diff
@@ -34,154 +37,223 @@ type State = AsyncComponent['state'] & {
...
+  align-items: center;
 `;
 
-const Name = styled('h4')`
-  font-size: ${p => p.theme.fontSizeLarge} !important;
-  font-weight: 600;
-  margin-bottom: 0 !important;
-  color: ${p => p.theme.gray600};
+const StyledButton = styled(Button)`
+  width: 40px;
```
Comment (matejminar) [inline src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx:253]: I would really love it if we could do these action buttons without additional styles. 
It looks like we are moving to this kind of design in settings more and more, so it would be great not having to override them on every page.

Do you know if passing size="small" to a button and size="sm" to an icon would be close enough to design?
I did it that way in Source Maps settings. @doralchan wdyt?

Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx:253]: sure..let me update it :)

File: src/sentry/static/sentry/app/views/settings/organizationRelays/modals/modalManager.tsx [hunk 21]
```diff
@@ -0,0 +1,228 @@
...
+class DialogManager<
+  P extends Props = Props,
+  S extends State = State
+> extends React.Component<P, S> {
+  constructor(props: P) {
+    super(props);
+    this.handleValidate = this.handleValidate.bind(this);
+    this.handleChange = this.handleChange.bind(this);
+    this.handleValidateKey = this.handleValidateKey.bind(this);
+    this.handleSave = this.handleSave.bind(this);
```
Comment (matejminar) [inline src/sentry/static/sentry/app/views/settings/organizationRelays/modals/modalManager.tsx:43]: Should we replace these with an arrow functions?

File: src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx [hunk 22]
```diff
@@ -34,154 +37,220 @@ type State = AsyncComponent['state'] & {
...
-  // TODO(Priscila): activate the code below as soon as the endpoint is provided
-  // getEndpoints(): ReturnType<AsyncComponent['getEndpoints']> {
-  //   return [['relays', `/organizations/${this.props.organization.slug}/relay-keys`]];
-  // }
+  setRelays(trustedRelays: Array<Relay>) {
+    this.setState({relays: trustedRelays});
+  }
 
-  // handleDelete = (id: Relay['public_key']) => () => {};
+  handleDelete(publicKey: Relay['publicKey']) {
```
Comment (matejminar) [inline src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx:57]: Do theese handle methods need to be arrow function too?

File: src/sentry/static/sentry/app/views/settings/organizationRelays/modals/modalManager.tsx (hunk 21 above)
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/settings/organizationRelays/modals/modalManager.tsx:43]: I'd love too, but I got a feedback from Evan saying that arrow funcs inside classes are much slower than bound funcs ...because of that I decided to convert them

File: src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx (hunk 22 above)
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx:57]: I decided to make only the return method an arrow function because it already has the bind

File: src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx [hunk 23]
```diff
@@ -34,154 +37,223 @@ type State = AsyncComponent['state'] & {
...
-  padding: ${space(1)} ${space(2)};
-  :last-child {
-    border-bottom: 0;
+const StyledPanelTable = styled(PanelTable)`
+  grid-template-columns: repeat(3, auto) max-content;
+  > * {
+    padding: ${space(1)};
+    @media (min-width: ${theme.breakpoints[0]}) {
+      padding: ${space(1)} ${space(2)};
+    }
```
[the remark this answers is not in the document]
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx]: I had changed the padding because of the size of the button, it seemed to have a lot of space..but now it doesn't make sense. I've updated the code ;-)

File: src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx [hunk 24]
```diff
@@ -34,154 +37,223 @@ type State = AsyncComponent['state'] & {
...
+            return (
+              <React.Fragment key={key}>
+                {description ? (
+                  <Name>
+                    <Text>{name}</Text>
+                    <QuestionTooltip position="top" size="sm" title={description} />
+                  </Name>
+                ) : (
+                  <Text>{name}</Text>
+                )}
```
[the remark this answers is not in the document]
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/settings/organizationRelays/relays.tsx]: hm. I didn't want to render an extra div in the dom, but I'm not too fussy...with the div it's fine too 😉

File: src/sentry/static/sentry/app/views/settings/organizationRelays/modals/modalManager.tsx (hunk 21 above)
Comment (matejminar) [reply src/sentry/static/sentry/app/views/settings/organizationRelays/modals/modalManager.tsx:43]: Do you mean this comment? https://github.com/getsentry/sentry/pull/19506#discussion_r444437158

Comment (billyvg) [reply src/sentry/static/sentry/app/views/settings/organizationRelays/modals/modalManager.tsx:43]: +1 to what @matejminar said, in that case there was no reason to bind `this`, it's unrelated to performance
